<launch>
  <arg name="ns" default="tracking_ros" />
  <arg name="device" default="cuda:0" />
  <arg name="mode" default="interactive" /> <!-- interactive, all, grounded, detic -->
  <arg name="num_slots" default="-1" />

  <!-- sam -->
  <arg name="model_type" default="vit_h_hq" /> <!-- vit_b, vit_l, vit_h ,vit_b_hq, vit_l_hq, vit_h_hq -->

  <arg name="compressed" default="false"/>
  <arg name="input_image" default="/kinect_head/rgb/image_rect_color"/>
  <arg name="decompressed_image" default="/$(arg ns)/decompressed_image"/>
  <arg name="input_point" default="/$(arg ns)/processed_image/screenpoint" />
  <arg name="input_bbox" default="/$(arg ns)/processed_image/screenrectangle" />
  <arg name="processed_image" default="/$(arg ns)/processed_image"/>
  <arg name="segmentation_image" default="/$(arg ns)/segmentation_image" />
  <arg name="prompt_image" default="/$(arg ns)/prompt_image" />
  <arg name="input_segmentation" default="/$(arg ns)/input_segmentation" />
  <arg name="output_segmentation" default="/$(arg ns)/segmentation" />


  <group ns="$(arg ns)">
    <!-- process image and repub -->
    <node if="$(arg compressed)" name="raw_image_transport" pkg="image_transport" type="republish" args="compressed">
      <remap from="in" to="$(arg input_image)"/>
      <remap from="out" to="$(arg processed_image)"/>
    </node>
    <node unless="$(arg compressed)" name="raw_image_transport" pkg="image_transport" type="republish" args="raw">
      <remap from="in" to="$(arg input_image)"/>
      <remap from="out" to="$(arg processed_image)"/>
    </node>

    <!-- sam node -->
    <node name="sam_node"
            pkg="tracking_ros" type="sam_node.py"
            output="screen" >
        <remap from="~input_image" to="$(arg input_image)" />
        <remap from="~output_image" to="$(arg prompt_image)" />
        <remap from="~input_point" to="$(arg input_point)" />
        <remap from="~input_bbox" to="$(arg input_bbox)" />
        <remap from="~segmentation" to="$(arg input_segmentation)" />
        <rosparam subst_value="true" >
            model_path: $(find tracking_ros)/trained_data/sam/sam_$(arg model_type).pth
            device: $(arg device)
            model_type: $(arg model_type)
            mode: $(arg mode)
            num_slots: $(arg num_slots)
        </rosparam>
    </node>

    <!-- tracking node -->
    <node name="cutie_node"
          pkg="tracking_ros" type="cutie_node.py"
          output="screen" >
        <remap from="~input_image" to="$(arg input_image)" />
        <remap from="~input_segmentation" to="$(arg input_segmentation)" />
        <remap from="~output_image" to="$(arg segmentation_image)" />
        <remap from="~segmentation" to="$(arg output_segmentation)" />
        <rosparam subst_value="true" >
            device: $(arg device)
            weights: $(find tracking_ros)/trained_data/cutie/cutie-base-mega.pth
        </rosparam>
    </node>

    <!-- interactive prompter -->
    <group if="$(eval mode=='interactive')">
      <node name="image_view2" pkg="image_view2" type="image_view2" output="screen" >
        <remap from="image" to="$(arg processed_image)" />
        <param name="use_window" value="false" />
      </node>
      <node  pkg="jsk_rqt_plugins" type="rqt_image_view2" name="prompter"
            output="screen">
        <remap from="event" to="image_view2/event" />
        <remap from="image_marked" to="image_marked" />
      </node>
      <node name="prompter_button"
            pkg="jsk_rqt_plugins" type="rqt_tabbed_buttons" output="screen" clear_params="true">
        <rosparam command="load" file="$(find tracking_ros)/resource/remote_tabbed_layout.yaml" />
      </node>

      <!-- visualize segmentation -->
      <node name="image_view" pkg="image_view" type="image_view" respawn="false" output="screen">
      <remap from="image" to="$(arg prompt_image)" />
      </node>
    </group>
  </group>
</launch>
